apps/aesophia/src/aeso_icode_to_asm.erl

```diff
@@ -426,7 +427,9 @@ assemble_expr(Funs, Stack, Tail,
     , dup(3)                                  %%  OOffset OSize ISize OOffset IOffset To Gas Value
     , shuffle_stack([1, 2, 4, 3, 5, 7, 8, 6]) %%  Gas To Value IOffset ISize OOffset OSize OOffset
     , i(?CALL)                                %%  Result OOffset
-    , pop(1)                                  %%  OOffset         -- ignoring result for now
+    , jump_if(CallOk) %% Jump if non-zero. CALL places on stack 1 if ok and 0 if failure. I.e. jump if non-failure. TODO: stricter check on ok (1) rather than non-failure.
+    , exhaust_gas()
```

Comment:
It seems to me that a better solution would be to have `CALL` exhaust the gas if not successful. Is there a use case where you'd want to recover from a non-successful call (what would the return data be?)?

Also note that I'm changing the semantics of `CALL` in [PT-160377912-cheap-encode-decode](https://github.com/aeternity/epoch/tree/PT-160377912-cheap-encode-decode) for Sophia contracts to return the data on the stack instead of a simple 0/1.

Reply:
> It seems to me that a better solution would be to have `CALL` exhaust the gas if not successful.

I agree, it should keep these changes re check on value smaller. I will revert&amend.

(I see in https://github.com/aeternity/epoch/tree/PT-160377912-cheap-encode-decode returning 0 **and** exhausting gas is done (`{0, aevm_eeevm_state:set_gas(0, ReturnState2)}`).)

> Is there a use case where you'd want to recover from a non-successful call (what would the return data be?)?

I would expect the return data to be dummy hence not meant to be used. I expect this to be more a question of level of compatibility among different VM ABIs. I consider these cases tracked as https://www.pivotaltracker.com/story/show/159686915

(I agree to your proposal above, will amend.)

> Also note that I'm changing the semantics of CALL ...

I see some changes in CALL semantics e.g. "In Sophia the ISize is the type of the arguments and IOffset is a pointer into the heap." . Thanks for the heads up.

Reply:
Addressed. CI is running.